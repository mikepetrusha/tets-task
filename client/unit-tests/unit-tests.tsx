/* eslint-disable import/export */
import { render, RenderOptions } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { RouterProvider } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { router } from '@/router';

import { GlobalStyle } from '../src/config/theme';
import theme from '../src/config/theme/theme';

const AllProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
    </ThemeProvider>
  );
};

export const TestBrowserProvider = () => {
  return <RouterProvider router={router} />;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
