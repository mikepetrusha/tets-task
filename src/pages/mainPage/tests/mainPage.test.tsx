import React from 'react';
import renderer from 'react-test-renderer';

import { fireEvent, render, screen, waitFor } from '../../../../unit-tests';
import MainPage from '../mainPage';

it('renders correctly', () => {
  const tree = renderer.create(<MainPage />).toJSON();
  expect(tree).toMatchSnapshot();
});

const mockData = [
  {
    id: '1',
    firstname: 'test1',
    lastname: 'test1',
    email: 'test1@gmail.com'
  },
  {
    id: '2',
    firstname: 'test2',
    lastname: 'test2',
    email: 'test2@gmail.com'
  },
  {
    id: '3',
    firstname: 'test3',
    lastname: 'test3',
    email: 'test3@gmail.com'
  }
];

const mockUseContactsService = jest.fn(() => ({
  getContacts: jest.fn(),
  contacts: mockData,
  isLoading: false
}));

jest.mock('@/services/contacts', () => ({
  useContactsService: () => mockUseContactsService()
}));

// Mocking useCreateModal and useEditModal hooks
jest.mock('@/components/create-modal/create-modal', () => ({
  useCreateModal: () => [jest.fn()]
}));

jest.mock('@/components/edit-modal/edit-modal', () => ({
  useEditModal: () => [jest.fn()]
}));

describe('Main page', () => {
  it('should render Cards', async () => {
    render(<MainPage />);
    await waitFor(() => {
      const contacts = screen.getAllByTestId('test-contact-card');
      expect(contacts).toHaveLength(3);
    });
  });

  it('should render "No data found" message when no contacts are available', async () => {
    mockUseContactsService.mockReturnValueOnce({
      getContacts: jest.fn(),
      contacts: [],
      isLoading: false
    });

    render(<MainPage />);

    const noDataMessage = screen.getByText('No data found');
    expect(noDataMessage).toBeInTheDocument();
  });

  it('should render Spinner while loading contacts', async () => {
    mockUseContactsService.mockReturnValueOnce({
      getContacts: jest.fn(),
      contacts: [],
      isLoading: true
    });

    render(<MainPage />);

    const spinner = screen.getByTestId('test-spinner');
    expect(spinner).toBeInTheDocument();
  });

  it('should open create modal when "NEUER EINTRAG" button is clicked', async () => {
    mockUseContactsService.mockReturnValueOnce({
      getContacts: jest.fn(),
      contacts: [],
      isLoading: false
    });

    render(<MainPage />);

    const createButton = screen.getByText('NEUER EINTRAG');
    fireEvent.click(createButton);
  });

  it('should open edit modal when a contact card is clicked', async () => {
    const mockContacts = [
      {
        id: '1',
        firstname: 'test1',
        lastname: 'test1',
        email: 'test1@gmail.com'
      },
      {
        id: '2',
        firstname: 'test2',
        lastname: 'test2',
        email: 'test2@gmail.com'
      },
      {
        id: '3',
        firstname: 'test3',
        lastname: 'test3',
        email: 'test3@gmail.com'
      }
    ];
    mockUseContactsService.mockReturnValueOnce({
      getContacts: jest.fn(),
      contacts: mockContacts,
      isLoading: false
    });

    render(<MainPage />);

    const contactCards = screen.getAllByTestId('test-contact-card');
    fireEvent.click(contactCards[0]);
  });
});
