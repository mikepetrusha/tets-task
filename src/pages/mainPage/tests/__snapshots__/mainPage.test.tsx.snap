// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-fHejqy cbXTTA"
>
  <div
    className="sc-hIPBNq iqogkL"
  >
    <button
      className="sc-hLQSwg euuwXi"
      onClick={[Function]}
    >
      NEUER EINTRAG
    </button>
  </div>
  <div
    className="sc-dmXWDj UnVpE"
  >
    <button
      className="sc-hLQSwg imjtzi"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="sc-blmEgr guBWWC"
  >
    <button
      className="sc-dJGMql GIEgC"
      onClick={[Function]}
    >
      <div
        className="sc-eDLKkx kbaZxH sc-ktwOfi gIAEGN"
        data-testid="test-contact-card"
      >
        <p
          className="sc-hABBmJ bguDKy"
        >
          test1
           
          test1
        </p>
        <p
          className="sc-fmKFGs jUWDXM"
        >
          test1@gmail.com
        </p>
      </div>
    </button>
    <button
      className="sc-dJGMql GIEgC"
      onClick={[Function]}
    >
      <div
        className="sc-eDLKkx kbaZxH sc-ktwOfi gIAEGN"
        data-testid="test-contact-card"
      >
        <p
          className="sc-hABBmJ bguDKy"
        >
          test2
           
          test2
        </p>
        <p
          className="sc-fmKFGs jUWDXM"
        >
          test2@gmail.com
        </p>
      </div>
    </button>
    <button
      className="sc-dJGMql GIEgC"
      onClick={[Function]}
    >
      <div
        className="sc-eDLKkx kbaZxH sc-ktwOfi gIAEGN"
        data-testid="test-contact-card"
      >
        <p
          className="sc-hABBmJ bguDKy"
        >
          test3
           
          test3
        </p>
        <p
          className="sc-fmKFGs jUWDXM"
        >
          test3@gmail.com
        </p>
      </div>
    </button>
  </div>
</div>
`;
